components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+    /// A list of one or more values.
+    List(Vec<ComputedValueItem>),
+}
+
+impl OneOrMoreSeparated for ComputedValueItem {
+    type S = Space;
+}
+
+/// A map from CSS variable names to CSS variable computed values, used for
+/// resolving.
+///
+/// This composes a custom_properties::CustomPropertiesMap, which maps CSS
+/// variable names to their token stream values.
+///
+/// We also keep track of whether or not this contains any uninherited
+/// properties, in order to just clone an Arc when possible.
+#[derive(Clone, Debug, PartialEq)]
+pub struct CustomPropertiesMap {
+    untyped_map: custom_properties::CustomPropertiesMap,
+    typed_map: HashMap<Name, ComputedValue>,
+    has_uninherited: bool,
+}
+
+impl CustomPropertiesMap {
+    /// Creates a new custom properties map.
+    pub fn new() -> Self {
+        CustomPropertiesMap {
+            untyped_map: custom_properties::CustomPropertiesMap::new(),
+            typed_map: Default::default(),
+            has_uninherited: false,
+        }
+    }
+
+    // We reuse almost everything from custom_properties::CustomPropertiesMap
+    // due to our Deref implementation.
+
+    /// Insert both an uncomputed and computed value for a typed custom property
+    /// if they have not previously been inserted. This must not be called for
+    /// untyped custom properties. The untyped custom property is not inserted
+    /// if it is not provided (currently this is only used so that
+    /// custom_properties::substitute_one doesn't need to be aware of typed
+    /// custom properties).
+    pub fn insert_typed(
+        &mut self,
+        name: &Name,
+        untyped_value: Option<custom_properties::ComputedValue>,
+        typed_value: ComputedValue,
+    ) {
+        debug_assert!(!self.typed_map.contains_key(name));
+        if let Some(untyped_value) = untyped_value {
+            self.untyped_map.insert(name.clone(), untyped_value);
+        }
+        self.typed_map.insert(name.clone(), typed_value);
+    }
+
+    /// Get the (typed) computed value.
+    /// It is not an error to call this with an untyped custom property: in that
+    /// case we just return None (callers should themselves check whether or not
+    /// a property has been registered).
+    pub fn get_typed(&self, name: &Name) -> Option<&ComputedValue> {
+        self.typed_map.get(name)
+    }
+
+    /// Mark that this map contains a non-initial value for an uninherited
+    /// custom property.
+    pub fn set_has_uninherited(&mut self) {
+        self.has_uninherited = true
+    }
+
+    /// Return whether this map contains a non-initial value for an uninherited
+    /// custom property.
+    pub fn has_uninherited(&self) -> bool {
+        self.has_uninherited
+    }
+}
+
+impl Deref for CustomPropertiesMap {
+    type Target = custom_properties::CustomPropertiesMap;
+
+    fn deref(&self) -> &Self::Target {
+        &self.untyped_map
+    }
+}
+
+impl DerefMut for CustomPropertiesMap {
```

Comment:
This looks like a bit of a hack to avoid changing more code :)

It's fine I suppose, but...